### PR TITLE
Correct attempt to instanciate isAbstract Class and look for sub@Type

### LINKED ISSUE #403: Enable serialization group in doctrine ORM yml file
Hello,

following the doc, to enable serialization group on property, it's needed to set annotations : 

```php
    /**
     * @Groups({"contacts"})
     */
    protected $moyensComm;
```

I would like to enable it using yml ORM doctrine file instead : 

```yml
ContactBundle\Entity\Ami:
    associationOverride:
        moyensComm:
            groups: [ "contacts" ]
```

It looks like adding serialization group inside yaml doctrine file has no effect.

I'm in need of this because i get error when I'm trying to add a new phone at Contact->Ami : 

Contact is a mappedSuperclass. Ami/Ennemi class extends that mappedSuperclass.
Contact got manyToMany association to abstract class moyensComm and Telephone class extends moyensComm (with alone 'numero' property).

Error is explicit, it tells to define serialization group on moyensComm Ami class property inherited from from Contact. SerializationGroup is already enabled on moyensComm Contact class property. But as moyensComm property do not exists in Ami/Ennemi class, and only a associationOverride is defined in these classes, how to set the serializationGroup in it ?

That's why I wanna use yml ORM Doctrine file instead of annotations to be able to define it inside associationOverride.

Thanks for help.

### PR DESCRIPTION
**Pull request by 3kynox**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #403 
| License       | MIT
| Doc PR        | N/A

All credits remains to @jderusse 

While denormalization, we use metadata defined in entity to generate instance and we mask the "@type" part.

It works but we can have some side effects both on normalizer or on ResourceResolver.

Behat test's coming.

Resolves #403

### CODE REVIEW

File: JsonLd/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -209,12 +209,19 @@ public function denormalize($data, $class, $format = null, array $context = [])
         $instanceClass = $overrideClass ? get_class($context['object_to_populate']) : $class;
         $reflectionClass = new \ReflectionClass($instanceClass);
         if ($reflectionClass->isAbstract()) {
-            throw new InvalidArgumentException(
-                sprintf(
-                    'Cannot create an instance of %s from serialized data because it is an abstract resource',
-                    $instanceClass
-                )
-            );
+            $isAbstract = true;
```
Comment (jderusse) [inline JsonLd/Serializer/ItemNormalizer.php]: Unused variable

File: JsonLd/Serializer/ItemNormalizer.php [hunk 2]
```diff
@@ -209,12 +209,19 @@ public function denormalize($data, $class, $format = null, array $context = [])
...
+            if (is_subclass_of($resource->getEntityClass(), $instanceClass)) {
+                $instanceClass = $resource->getEntityClass();
+                $reflectionClass = new \ReflectionClass($instanceClass);
```
Comment (jderusse) [inline JsonLd/Serializer/ItemNormalizer.php]: Unused variable

File: JsonLd/Serializer/ItemNormalizer.php [hunk 3]
```diff
@@ -209,12 +209,19 @@ public function denormalize($data, $class, $format = null, array $context = [])
...
         if ($reflectionClass->isAbstract()) {
-            throw new InvalidArgumentException(
-                sprintf(
-                    'Cannot create an instance of %s from serialized data because it is an abstract resource',
-                    $instanceClass
-                )
-            );
+            $isAbstract = true;
+            if (is_subclass_of($resource->getEntityClass(), $instanceClass)) {
+                $instanceClass = $resource->getEntityClass();
```
Comment (jderusse) [inline JsonLd/Serializer/ItemNormalizer.php]: You have to check if the instanceClass is abstract or not and throw an exception if it is.

Comment (jderusse) [inline JsonLd/Serializer/ItemNormalizer.php]: I wonder if we shouldn't use the class defined in the resource every time and not only when the class passed to the denormalized function is abdtarct.

Any advice @dunglas ?

**Comment by 3kynox** [comment]
Corrected, big thanks to you @jderusse

File: Api/ResourceResolverTrait.php [hunk 4]
```diff
@@ -54,7 +54,9 @@ public function guessResource($object, array $context = null, $strict = false)
             $type = gettype($type);
         }
 
-        if (isset($context['resource'])) {
+        if (is_array($object) && isset($object['@type'])) {
```
Comment (dunglas) [inline Api/ResourceResolverTrait.php]: It cannot work because `@type` can be any IRI like `http://schema.org/Product`. And it's why tests fail.

**Comment by dunglas** [comment]
The only way I can see to do cleanly what you want is to maintain a correspondance table between JSON-LD `@type` and PHP classes.
This data is available in Resources metadata, but IMO it's a change too big for v1. It should be done in the master branch.

I'm not a sure such an edge case deserve the introduction of this complexity. Why don't you create your related entities one by one doing POST or PUT request on their own collections, then deal directly with IRIs?

**Comment by 3kynox** [comment]
Thanks @dunglas for review.

Simply because it force to use more than one request.
If I take the case I need to create multiple subTypes from one SPA form, too much requests will have to be done, and it will decrease performance of application built with Api-Platform.

I think it's better to include any mapping cases that is possible with Doctrine with a global api system, such as api-platform.

PR have to be re-done on master ?
To do things in a more cleaned way, my knowledge won't be suffisant at this time, and hope some help could be acquired.

**Comment by benjaminrau** [comment]
@3kynox @dunglas The use case which is discussed here fits exactly my one. In my application it is possible to create related entities with separate api call for a performance point of view. I would really appriciate this fix comes to the master. 

Anyway @dunglas is right that you cant lookup the resource by type property because it can be a IRI from schema.org fx. So we would need to lookup resource reading IRI from annotations?

**Comment by dunglas** [comment]
Yes, more precisely, the IRI value is extracted form annotations. XML or YAML config and stored in metadata. The lookup should be done from metadata and fallback to the default name. 

Anybody interested to work on this?

**Comment by dunglas** [comment]
Closing because it's not possible to that this way. See also #663 and #615.

**Pull request closed without merging**